extensions/amp-fx-collection/0.1/providers/parallax.js

```diff
@@ -0,0 +1,187 @@
+   */
+  installOn(element) {
+    const parallaxElement = new ParallaxElement(
+        element, this.positionObserver_, this.viewport_, this.vsync_);
+    parallaxElement.initialize();
+  }
+}
+
+/**
+ * Encapsulates and tracks an element's linear parallax effect.
+ */
+class ParallaxElement {
+  /**
+   * @param {!Element} element The element to give a parallax effect.
+   * @param {!../../../../src/service/position-observer/position-observer-impl.PositionObserver} positionObserver
+   * @param {!../../../../src/service/viewport/viewport-impl.Viewport} viewport
+   * @param {!../../../../src/service/vsync-impl.Vsync} vsync
+   */
+  constructor(element, positionObserver, viewport, vsync) {
+
+    /** @private @const {!../../../../src/service/position-observer/position-observer-impl.PositionObserver} */
+    this.positionObserver_ = positionObserver;
+
+    /** @private @const {!../../../../src/service/viewport/viewport-impl.Viewport} */
+    this.viewport_ = viewport;
+
+    /** @const @private {!../../../../src/service/vsync-impl.Vsync} */
+    this.vsync_ = vsync;
+
+    /** @const {string} */
+    const factorValue = user().assert(element.getAttribute(FACTOR_ATTR),
+        `${FACTOR_ATTR}=<number> attribute must be provided for: %s`, element);
+
+    /** @private @const {number} */
+    this.factor_ = parseFloat(factorValue);
+
+    /** @private {?number} */
+    this.adjustedViewportHeight_ = null;
+
+    user().assert(this.factor_ > 0,
+        `${FACTOR_ATTR} must be a number and greater than 0 for: %s`, element);
+
+    /** @private @const {!Element} */
+    this.element_ = element;
+  }
+
+  /**
+   * Handles initializations such as getting initial positions and listening to
+   * events.
+   */
+  initialize() {
+    this.getAdjustedViewportHeight_().then(adjustedViewportHeight => {
+      this.adjustedViewportHeight_ = adjustedViewportHeight;
+
+      // start observing position of the element.
+      this.observePositionChanges_();
+    });
+  }
+
+  /**
+   * Apply the parallax effect to the offset given how much the page
+   * has moved since the last frame.
+   * @private
+   */
+  update_(entry) {
+    dev().assert(this.adjustedViewportHeight_);
+    // outside viewport
+    if (!entry.positionRect ||
+        entry.positionRect.top > this.adjustedViewportHeight_) {
+      return;
+    }
+
+    // User provided factor is 1-based for easier understanding.
+    // Also negating number since we are using tranformY so negative = upward,
+    // positive = downward.
+    const adjustedFactor = -(this.factor_ - 1);
+    const top = entry.positionRect.top;
+    // Offset is how much extra to move the element which is position within
+    // viewport times adjusted factor.
+    const offset = (this.adjustedViewportHeight_ - top) * adjustedFactor;
+
+    // Translate the element offset pixels.
+    // No need for vsync mutate, position observer only calls back at most
+    // every animation frame.
+    setStyles(this.element_,
```

Comment:
I thought the purpose of vsync mutate was not to batch individual components' calls, but to batch mutations across all components that might manipulate the DOM at any moment? Wouldn't you still need it here to coordinate with other AMP components manipulating the DOM? I'm still trying to learn when it is and isn't needed 😅

Reply:
Your assessment is correct but we still don't need it here since we are only changing `transform` which is done at composition layer and won't cause relayouts. Updated the comments (it also reminded me to add `'will-change': 'transform'` to the element as well) . Using `vsyncMutate` would make us be one frame behind which doesn't feel as nice.

Reply:
kk thanks for explaining